clesperanto/core/plugin_function.py

```diff
@@ -0,0 +1,79 @@
+import inspect
+from typing import Any, Callable, Dict, Optional, Sequence, Set, Type, Union
+from functools import wraps
+from toolz import curry
+
+from gputools import OCLArray
+
+from .create import create_like
+from .types import Image, isImage
+from .push import push
+
+@curry
+def plugin_function(
+    function: Callable,
+    output_creator: Callable = create_like
+) -> Callable:
+    """Function decorator which ensures correct types and values of parameters
+        Given input parameters are either of type OCLArray (which the GPU understands) or are converted to this type
+        (see push function). If output parameters of type OCLArray are not set, an empty image is created and handed
+        over.
+
+    :param function: the function to be called on the GPU
+    :param output_creator: a function which can create an output OCLArray given an input OCLArray, optional
+        per default, we create output images of the same shape as input images
+    :return: return value of the executed function
+    """
+
+
+    @wraps(function)
+    def worker_function(*args, **kwargs):
+        # determine argument spec and default values, values are given as args
+        argument_specification = inspect.getfullargspec(function)
+        defaults_values = argument_specification.defaults
+
+        target_arguments = {} # empty dictionary to store parameters as we forward them
+
+        default_counter = 0
+        any_ocl_input = None
+
+        for arg_counter, argument in enumerate(argument_specification.args):
+            #print("---\nparsing argument " + argument)
+            if (len(args) > arg_counter):
+                value = args[arg_counter]
+            else:
+                value = None
```

Comment:
Are we missing kwargs here?

Reply:
ood catch!
 I'm combining all parameters in kwargs now. This obviously includes the given ones